### PR TITLE
WIP - aragon-voting-rewards

### PR DESCRIPTION
**Pull request by allemanfredi**
I'm opening this PR in order to review what have been until now.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# :moneybag: voting-reward
+# :moneybag: voting-rewards
 
 An Aragon app that allows to get rewards in base of how many votes you made in a certain amount of time.
```
[the remark this answers is not in the document]
Comment (allemanfredi) [reply README.md:3]: solved

File: package.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "voting-reward",
+  "name": "voting-rewards",
   "version": "0.0.1",
   "description": "An Aragon app that allows to get rewards in base of how many votes you made in a certain amount of time",
```
[the remark this answers is not in the document]
Comment (allemanfredi) [reply package.json:4]: solved

File: contracts/VotingRewards.sol [hunk 3]
```diff
@@ -395,13 +410,13 @@ contract VotingReward is AragonApp {
      * @param _beneficiary beneficiary
      * @param _from block from wich starting looking for votes
      * @param _to   block to wich stopping looking for votes
-     * @param _missingVotesThreeshold number of vote to which is possible to don't vote
+     * @param _missingVotesThreshold number of vote to which is possible to don't vote
```
[the remark this answers is not in the document]
Comment (allemanfredi) [reply contracts/VotingRewards.sol:436]: solved

File: contracts/VotingRewards.sol [hunk 4]
```diff
@@ -395,13 +410,13 @@ contract VotingReward is AragonApp {
      * @param _beneficiary beneficiary
      * @param _from block from wich starting looking for votes
```
[the remark this answers is not in the document]
Comment (allemanfredi) [reply contracts/VotingRewards.sol:434]: solved

File: contracts/VotingRewards.sol [hunk 5]
```diff
@@ -226,16 +229,16 @@ contract VotingReward is AragonApp {
 
     /**
      * @notice Change minimum number of missing votes allowed
-     * @param _missingVotesThreeshold number of seconds minimun to claim access to voting rewards
+     * @param _missingVotesThreshold number of seconds minimun to claim access to voting rewards
```
[the remark this answers is not in the document]
Comment (allemanfredi) [reply contracts/VotingRewards.sol]: solved

File: contracts/VotingRewards.sol [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+        uint64 _fromBlock,
+        uint64 _toBlock,
+        uint256 _missingVotesThreshold
+    ) internal view returns (uint256) {
+        uint256 missingVotes = 0;
+        uint256 minimunBalance = 0;
+        bool forceFirstAssignment = true;
+
+        // voteId starts from 1 in DandelionVoting
+        for (
```
Comment (gskapka) [inline contracts/VotingRewards.sol:457]: Now that you've updated the loop in `_calculateReward`, why add 1 to `votes.length` then immediately subtract it again in the loop to get the vote ID?

Just use the following loop and get rid of `realVoteId` entirely:

```
for (uint256 voteId = dandelionVoting.votesLength(); voteId >= 1; voteId--) {
    ...
}
```

File: contracts/VotingRewards.sol [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+                    .getVoterState(realVoteId, _beneficiary);
+
+                if (state == DandelionVoting.VoterState.Absent) {
+                    missingVotes = missingVotes.add(1);
+                }
+
+                require(
+                    missingVotes <= _missingVotesThreshold,
+                    ERROR_TOO_MANY_MISSING_VOTES
+                );
```
Comment (gskapka) [inline contracts/VotingRewards.sol:479]: This worries me - having this `require` in here proves that this function is not tested enough, as if it was one of your tests would ought now to be failing.

If a user did not miss voting on the first vote checked on the first turn of this loop, `missingVotes` here will **always** be 0 and thus **always** be failing.  

In fact, if the `_missingVoteThreshold` is anything > 1, we can _never_ pass this require statement.

File: contracts/VotingRewards.sol [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+
+                uint256 votingTokenBalanceAtVote = MiniMeToken(
+                    dandelionVoting.token()
+                )
+                    .balanceOfAt(_beneficiary, startBlock);
+
+                if (forceFirstAssignment == true) {
+                    forceFirstAssignment = false;
+                    minimunBalance = votingTokenBalanceAtVote;
+                }
```
Comment (gskapka) [inline contracts/VotingRewards.sol:489]: This is wasteful to have the extra bool here to check first assignment. Instead, make a variable of the `votes.length()`, then use that variable in the loop declaration _and_ here in order to check for the first turn of the loop (ie `if (numberOfVotes == voteId) { /* make first assignment */ }`

File: contracts/VotingRewards.sol [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+                if (votingTokenBalanceAtVote < minimunBalance) {
+                    minimunBalance = votingTokenBalanceAtVote;
+                }
+            }
+
+            // NOTE: avoid "out of epoch" cycles
+            if (startBlock < _fromBlock) break;
+        }
+
+        return (minimunBalance).mul(percentageRewards).div(PCT_BASE);
```
Comment (gskapka) [inline contracts/VotingRewards.sol:500]: Is it worth checking for `minimumBalance` of 0 here first, and just returning that if that's the case, and so only perform the calculation if `minimumBalance` > 0?

File: contracts/VotingRewards.sol [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+        emit RewardDistributed(_beneficiary, rewardAmount, lockTime);
+    }
+
+    /**
+     * @notice collect rewards for an address if lockTime is passed since when tokens have been distributed
+     * @param _beneficiary address that should be fund with rewards
+     * @dev rewardsVault should have TRANSFER_ROLE permission
+     */
+    function collectRewardsFor(address _beneficiary) public {
+        uint64 timestamp = getBlockNumber64();
```
Comment (gskapka) [inline contracts/VotingRewards.sol:410]: This is still _not at timestamp_. It's a _blockNumber_.

File: contracts/VotingRewards.sol [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+                    _beneficiary,
+                    rewards[i].amount
+                );
+                collectedRewards = collectedRewards.add(1);
+
+                emit RewardCollected(_beneficiary, rewards[i].amount);
+            }
+        }
+
+        require(collectedRewards >= 1, ERROR_NO_REWARDS);
```
Comment (gskapka) [inline contracts/VotingRewards.sol:434]: Is it worth returning `true` from this function?

File: contracts/VotingRewards.sol (hunk 6 above)
Comment (allemanfredi) [reply contracts/VotingRewards.sol:457]: yeah! correct!

**Comment by allemanfredi** [comment]
Just checked new the stuff and i noticed that `voteId == votesLength` is not a good condition because it could happen that a user did not vote at the first vote of an epoch but to all the others and therefore the user would not take the reward
although it is eligible (since has voted to all votes except threshold = 1)

**Comment by gskapka** [comment]
Does `_calculateReward` not just get more and more expensive over time as a user votes more and more? I don't know enough about the `DandelionVoting` stuff you're using - are votes ever deleted?

**Comment by allemanfredi** [comment]
Yes you are right! but since I start cycling from the last one and because are not deleted, this guarantees me that I will first access the votes relating to the last epoch.

File: package.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "voting-rewards",
-  "version": "0.0.1",
+  "version": "0.0.3",
```
Comment (gskapka) [inline package.json:3]: The typical order of the versioning is `major.minor.bugfix`, where major is something that's a breaking change. 

So this PR likely want's the minor version number bumping, not the bugfix one.

Comment (allemanfredi) [reply package.json:3]: yesterday i had to fix a small bug for a testnet deployment so it should be fine!

Comment (gskapka) [reply package.json:3]: Ah fair enough. Though this was for this whole PR, nevermind, carry on!

**Comment by gskapka** [comment]
Merging unreviewed per @allemanfredi 's request.

**Pull request merged**